packages/core/src/api/positionMapping.ts

```diff
@@ -0,0 +1,145 @@
+import type { Transaction } from "prosemirror-state";
+import { Mapping } from "prosemirror-transform";
+import {
+  absolutePositionToRelativePosition,
+  relativePositionToAbsolutePosition,
+  ySyncPluginKey,
+} from "y-prosemirror";
+import type { BlockNoteEditor } from "../editor/BlockNoteEditor.js";
+import * as Y from "yjs";
+import type { ProsemirrorBinding } from "y-prosemirror";
+import type {
+  DefaultInlineContentSchema,
+  DefaultStyleSchema,
+} from "../blocks/defaultBlocks.js";
+import type { DefaultBlockSchema } from "../blocks/defaultBlocks.js";
+import type {
+  BlockSchema,
+  InlineContentSchema,
+  StyleSchema,
+} from "../schema/index.js";
+
+export function isRemoteTransaction(tr: Transaction) {
+  return tr.getMeta(ySyncPluginKey) !== undefined;
+}
+
+type YSyncPluginState = {
+  doc: Y.Doc;
+  binding: Pick<ProsemirrorBinding, "type" | "mapping">;
+};
+
+/**
+ * This is used to keep track of positions of elements in the editor.
+ * It is needed because y-prosemirror's sync plugin can disrupt normal prosemirror position mapping.
+ *
+ * It is specifically made to be able to be used whether the editor is being used in a collaboratively, or single user, providing the same API.
+ */
+export class PositionStorage<
+  BSchema extends BlockSchema = DefaultBlockSchema,
+  ISchema extends InlineContentSchema = DefaultInlineContentSchema,
+  SSchema extends StyleSchema = DefaultStyleSchema
+> {
+  private readonly editor: BlockNoteEditor<BSchema, ISchema, SSchema>;
+  /**
+   * Whether the editor has had a remote transaction.
+   */
+  private hadRemoteTransaction = false;
+
+  private readonly mapping = new Mapping();
+  private mappingLength = 0;
+
+  constructor(
+    editor: BlockNoteEditor<BSchema, ISchema, SSchema>,
+    { shouldMount = true }: { shouldMount?: boolean } = {}
+  ) {
+    this.editor = editor;
+    this.onTransactionHandler = this.onTransactionHandler.bind(this);
+
+    if (!shouldMount) {
+      return;
+    }
+
+    if (!this.editor._tiptapEditor) {
+      throw new Error("Editor not mounted");
+    }
+    this.editor._tiptapEditor.on("transaction", this.onTransactionHandler);
+  }
```

Comment:
An alternative to this, would be to have a prosemirror plugin that gets these transactions between state updates. Perhaps that is preferable?

Reply:
this is fine imo, unless there's an advantage to doing this in a plugin, but I don't think so.

Another alternative would be to emit our own event in BlockNoteEditor.dispatch - but there's a change that misses things when people use _tiptapEditor.dispatch